docs/design/2018-09-04-histograms-in-plan.md

```diff
@@ -0,0 +1,116 @@
+# Proposal: Maintain statistics in `Plan`
+
+- Author:     [Yiding CUI](https://github.com/winoros)
+- Last updated:  2018-09-04
+
+## Abstract
+
+This proposal proposes to maintain the histogram in `Plan`’s statistics information.
+
+
+## Background
+
+Currently, TiDB only uses statistics when deciding which physical scan method a table should use. And TiDB only stores simple statistics in the plan structure. But when deciding the join order and considering some other optimization rules, we need more detailed statistics.
+
+So we need to maintain the statistics in the plan structure to get sufficient statistics information to do optimizations.
+
+## Proposal
+
+The new `statsInfo` of `plan` should be something like the following structure:
+
+```
+type statsInfo struct {
+	// This two attributes are already used in current statsInfo.
+	count float64
+	// The number of distinct value of each column.
+	ndv []float64
+
+	// HistColl collects histograms of columns.
+	histColl statistics.HistColl
+	// rangesOfXXX This won’t be maintained in the first version.
+	rangesOfColumn map[int][]*ranger.Range 
+	rangesOfIndex map[int][]*ranger.Range
+	// max/minValues is a simplified version of rangesOfXXX. Since maintaining rangesOfXXXX consumes a lot of time and memory, we can first check whether this can meet our needs.
+	maxValues map[int][]types.Datum
+	minValues map[int][]types.Datum
+}
+```
+
+This struct will be maintained when we call `deriveStats`.
+
+Currently we don't change the histogram itself during planning. Because it will consume a lot of time and memory space. I’ll try to maintain ranges slice or the max/min value to improve the accuracy of row count estimation instead.
+
+Before we maintain `rangesOfXXX` or `max and min values`:
+
+We maintain the histogram in `Projection`, `Selection`, `Join`, `Aggregation`, `Sort`, `Limit` and `DataSource` operators.
+
+For `Sort`, we can just copy children's `statsInfo` without doing any change.
+
+For `Limit`, we can just copy children's `statsInfo` or ignore the histogram information. As you know, its execution logic is based on randomization. It's hard to maintain the statistics information after `Limit`. But we may use the information before it to do some estimation in some scenarios.
+
+For `Projection`, change the reflection of the map we maintained.
+
+For `Aggregation`, use the histogram to estimate the NDV of group-by items. If one index cannot cover the group-by item, we’ll multiply the NDV of each group-by column. If the output of `Aggregation` includes group-by columns, we’ll maintain the histogram of them for future use.
```

Comment:
how to maintain the statistics info for the `Aggregate` operator?

Reply:
It may not need to do anything. Just use the child is okay.